### PR TITLE
Add taxes to undiscounted order prices

### LINKED ISSUE #12185: [RFC] Include taxes in `Order` undiscounted prices
### Problem
The `undiscounted` prices don't include the tax rate, the `undiscounted net` and `undiscounted gross` price is the same.
The return prices should include the same tax rate as the discounted prices.

### General Assumptions
- If the prices are entered as gross (with tax included), calculate each line's undiscounted net using undiscounted gross and the tax rate
- If the prices are entered as net (without tax included),  calculate each line's undiscounted gross using undiscounted net and the tax rate

### ToDo
- Update `complete_checkout` method to create `OrderLines` with correct `undiscounted prices` that include the tax rate (tax calculation should be done based on the `tax rate`, in similar way as it's done for a flat rates).
- Ensure that `create_order_from_checkout` method also calculated the correct `undiscounted prices`.
- Make sure subtotals and totals are correct.
- Update `fetch_order_prices_if_expired`mwthod - the undiscounted values should include taxes
- Ensure the `Order` `API` fields return the correct values for discounted and undiscounted fields
- Ensure that correct values are returned in the `Order` payload
- Add data migration for `OrderLine` to update the `undiscounted` prices to include the tax.

### PR DESCRIPTION
**Pull request by jakubkuc**
I want to merge this change because it fixes #12185

It adds calculating taxes for undiscounted prices for `OrderLine`

It contains:
- fix that will make new order lines have fields `undiscounted_unit_price` and `undiscounted_total_price` properly calculated gross and net values.
- Zero downtime migration with celery task that calculates and updates all OrderLines that have `tax_rate` > 0 and equal net and gross undiscounted prices.  

Acceptance criteria: 
- All queries For orders should contain proper `undiscounted_unit_price` and `undiscounted_total_price` fields 
- All new created and updated orders should also have  `undiscounted_unit_price` and `undiscounted_total_price` calculated correctly. 
 

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/tax/calculations/__init__.py [hunk 1]
```diff
@@ -18,3 +20,22 @@ def calculate_flat_rate_tax(
     return TaxedMoney(
         net=Money(net_amount, currency), gross=Money(gross_amount, currency)
     )
+
+
+def add_tax_to_undiscounted_price(
+    price: "Money", tax_rate: "Decimal", prices_entered_with_tax: bool
+) -> TaxedMoney:
```
Comment (IKarbowiak) [inline saleor/tax/calculations/__init__.py:27]: This method is almost the same as the above `calculate_flat_rate_tax`. IMO we should reuse the above one

File: saleor/order/migrations/tasks/saleor3_9.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    field_name, line, tax_rate, prices_entered_with_tax, lines_to_update
+):
+    line_net = getattr(line, f"{field_name}_net_amount")
+    line_gross = getattr(line, f"{field_name}_gross_amount")
+    if line_net == line_gross and all([line_net != 0, line_gross != 0]):
+        line_net, line_gross = add_taxes_to_price(
+            line_net, tax_rate, prices_entered_with_tax
+        )
+        setattr(line, f"{field_name}_net_amount", line_net)
+        setattr(line, f"{field_name}_gross_amount", line_gross)
```
Comment (IKarbowiak) [inline saleor/order/migrations/tasks/saleor3_9.py:39]: Shouldn't we quantize price value before saving?

File: saleor/checkout/complete_checkout.py [hunk 3]
```diff
@@ -249,6 +245,18 @@ def _create_line_for_order(
...
+        tax_rate=tax_rate,
+        prices_entered_with_tax=prices_entered_with_tax,
+    )
+
+    undiscounted_total_price = add_tax_to_undiscounted_price(
+        price=undiscounted_base_total_price,
+        tax_rate=tax_rate,
+        prices_entered_with_tax=prices_entered_with_tax,
+    )
+
```
Comment (IKarbowiak) [inline saleor/checkout/complete_checkout.py:259]: We need a test that is checking if a `undiscounted` price is with tax right now. Also, we need to ensure that for `create_order_from_checkout` the undiscounted prices are calculated with taxes too.

File: saleor/order/migrations/tasks/saleor3_9.py (hunk 2 above)
Comment (jakubkuc) [reply saleor/order/migrations/tasks/saleor3_9.py]: yes. fixed

**Review by IKarbowiak — COMMENTED** [review]
Please also handle the point from the RFC:
`Ensure that correct values are returned in the Order payload`

File: saleor/order/calculations.py [hunk 4]
```diff
@@ -44,19 +48,27 @@ def _recalculate_order_prices(
...
                 line.tax_rate = manager.get_order_line_tax_rate(
                     order, product, variant, None, line_unit.undiscounted_price
                 )
+                line.undiscounted_unit_price = add_tax_to_undiscounted_price(
+                    price=line.undiscounted_base_unit_price,
+                    tax_rate=line.tax_rate,
+                    prices_entered_with_tax=prices_entered_with_tax,
+                )
+                line.undiscounted_total_price = add_tax_to_undiscounted_price(
+                    price=line_total.undiscounted_price.net,
```
Comment (maarcingebala) [inline saleor/order/calculations.py:68]: We set price from `line_total.undiscounted_price.net`, which is taxed money, so it doesn't have the tax already?

**Review by IKarbowiak — COMMENTED** [review]
👮‍♀️  Please also handle the point from the RFC:
Ensure that correct values are returned in the Order payload

File: saleor/order/calculations.py [hunk 5]
```diff
@@ -44,19 +48,27 @@ def _recalculate_order_prices(
...
+                line.undiscounted_unit_price = calculate_flat_rate_tax(
+                    money=line.undiscounted_base_unit_price,
+                    tax_rate=line.tax_rate * 100,
+                    prices_entered_with_tax=prices_entered_with_tax,
+                )
+                line.undiscounted_total_price = calculate_flat_rate_tax(
+                    money=line_total.undiscounted_price.net,
+                    tax_rate=line.tax_rate * 100,
+                    prices_entered_with_tax=prices_entered_with_tax,
+                )
```
Comment (fowczarek) [inline saleor/order/calculations.py:71]: Should we `quantitize` those prices? `calculate_flat_rate_tax` doesn't have `quantize_price` inside.

File: saleor/order/migrations/tasks/saleor3_9.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    return net_amount, gross_amount
+
+
+def update_fields_for_line(
+    field_name, line, tax_rate, prices_entered_with_tax, lines_to_update
+):
+    line_net = getattr(line, f"{field_name}_net_amount")
+    line_gross = getattr(line, f"{field_name}_gross_amount")
+    if line_net == line_gross and all([line_net != 0, line_gross != 0]):
+        line_net, line_gross = add_taxes_to_price(
```
[the remark this answers is not in the document]
Comment (jakubkuc) [reply saleor/order/migrations/tasks/saleor3_9.py:35]: it is quantized later in `setattr`

File: saleor/checkout/complete_checkout.py [hunk 7]
```diff
@@ -249,6 +245,18 @@ def _create_line_for_order(
...
+        tax_rate=tax_rate * 100,
+        prices_entered_with_tax=prices_entered_with_tax,
+    )
+
+    undiscounted_total_price = calculate_flat_rate_tax(
+        money=undiscounted_base_total_price,
+        tax_rate=tax_rate * 100,
+        prices_entered_with_tax=prices_entered_with_tax,
+    )
+
```
[the remark this answers is not in the document]
Comment (jakubkuc) [reply saleor/checkout/complete_checkout.py:259]: yes. It was done with other function before and quantization was skipped when it was dropped during the development of this PR. Thanks, good catch!

File: saleor/order/calculations.py (hunk 5 above)
Comment (jakubkuc) [reply saleor/order/calculations.py:71]: yes. It was done with other function before and quantization was skipped when it was dropped during the development of this PR. Thanks, good catch!

**Review by fowczarek — APPROVED** [review]

File: saleor/order/migrations/tasks/saleor3_9.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    )
+    tax_configurations = tax_configuration.objects.filter(
+        Exists(channels.filter(tax_configuration__id="id"))
+    )
+    tax_conf_to_prices_entered_with_taxes = {
+        tc_id: price_entered_with_taxes
+        for tc_id, price_entered_with_taxes in tax_configurations.values_list(
+            "id", "prices_entered_with_tax"
+        )
+    }
```
Comment (maarcingebala) [inline saleor/order/migrations/tasks/saleor3_9.py:75]: If I'm correct, you could do:
```
dict(tax_configurations.values_list("id", "prices_entered_with_tax"))
```
But it's a stylistic change so not a big deal.

File: saleor/checkout/utils.py [hunk 9]
```diff
@@ -918,3 +919,26 @@ def get_external_shipping_id(container: Union["Checkout", "Order"]):
...
+):
+    """Apply taxes to undiscounted base price.
+
+    This function also prevents rounding difference between prices from tax-app and
+    local calculations based on tax_rate that might occur in orders without discounts.
+    """
+    if undiscounted_base_price == price.net and not prices_entered_with_tax:
+        return price
+    if undiscounted_base_price == price.gross and prices_entered_with_tax:
+        return price
```
Comment (fowczarek) [inline saleor/checkout/utils.py:936]:
```suggestion
    amount = price.net if not prices_entered_with_tax else price.gross
    if undiscounted_base_price == amount:
        return price
```

File: saleor/order/migrations/tasks/saleor3_9.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    order_to_prices_with_tax_map = get_order_prices_entered_with_tax_mapping(
+        order_lines,
+        order_model=Order,
+        channel_model=Channel,
+        tax_configuration=TaxConfiguration,
+    )
+
+    lines_to_update = set()
+    for line in order_lines:
+        prices_entered_with_tax = order_to_prices_with_tax_map[line.order.id]
```
Comment (korycins) [inline saleor/order/migrations/tasks/saleor3_9.py:119]: Have you tested the celery task with database min 1 mln of orders?
Also, you're calling`line.order.id` without prefetch.

**Review by korycins — COMMENTED** [review]
What about `undiscounted_total` on the Order? Don't we need to update it also? It is available via API.

File: saleor/order/migrations/tasks/saleor3_9.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+from ....core.prices import quantize_price
+from ....tax.models import TaxConfiguration
+from ....channel.models import Channel
+from ....order.models import OrderLine, Order
+from ....celeryconf import app
+
+task_logger = get_task_logger(__name__)
+
+BATCH_SIZE = 500
```
Comment (8r2y5) [inline saleor/order/migrations/tasks/saleor3_9.py:13]: Please do not use magic numbers, please test memory usage

**Comment by szczecha** [comment]
Hint for testing edge case with rounding issue: 

Sale inactive or removed 
Taxes
```
"channel":
"id":"Q2hhbm5lbDox"
"name":"Channel-USD"
"__typename":"Channel"
}
"displayGrossPrices":true
"pricesEnteredWithTax":false
"chargeTaxes":true
"taxCalculationStrategy":"TAX_APP"
```
Avatax plugin configured and activated in default-channel 
Product price netto 35 USD
Checkout create 
```
{
  "checkoutInput": {
    "channel": "default-channel",
        "email": "dustin.gonzalez@example.com",
        "lines": [
      {
        "quantity": 1,
        "variantId": "UHJvZHVjdFZhcmlhbnQ6MTcy"
      }
    ],

    "shippingAddress": {
        "country": "US",
        "city": "San Diego",
       "postalCode": "92128",
        "streetAddress1": "1852 Hood Avenue",
        "firstName": "jon",
        "lastName": "doe",
        "companyName": "xyz",
        "streetAddress2": "123abc",
        "countryArea": "California",
        "phone": "+48123456789"
    },
    "billingAddress": {
        "country": "US",
        "city": "San Diego",
       "postalCode": "92128",
        "streetAddress1": "1852 Hood Avenue",
        "firstName": "jon",
        "lastName": "doe",
        "companyName": "xyz",
        "streetAddress2": "123abc",
        "countryArea": "California",
        "phone": "+48123456789"
    }

  }
}
```

checkoutDeliveryMethodUpdate, use ID of first available shipping method returned for checkout
checkoutPaymentCreate plugin mirumee
checkoutComplete


It creates below order. If you look at order.lines.unitPrice and order.lines.undiscountedUnitPrice gross and tax have diffrent values
```

"data": {
        "checkoutComplete": {
            "order": {
                "id": "T3JkZXI6MTMwZjBhZmEtM2I0Yi00ZWMyLTliOTEtYzZlNWU2OTVjZjM5",
                "number": "36",
                "status": "UNFULFILLED",
                "paymentStatus": "FULLY_CHARGED",
                "total": {
                    "gross": {
                        "amount": 43.81,
                        "currency": "USD"
                    },
                    "net": {
                        "amount": 40.66,
                        "currency": "USD"
                    },
                    "tax": {
                        "amount": 3.15,
                        "currency": "USD"
                    }
                },
                "lines": [
                    {
                        "id": "T3JkZXJMaW5lOjU0MDM4YjBmLTZmYjctNDZlZS1hMTRhLTZmNDU2YWYzNjg1Ng==",
                        "productName": "Nebula Night Sky Paint",
                        "taxRate": 0.0775,
                        "totalPrice": {
                            "gross": {
                                "amount": 37.72,
                                "currency": "USD"
                            },
                            "net": {
                                "amount": 35.0,
                                "currency": "USD"
                            },
                            "tax": {
                                "amount": 2.72,
                                "currency": "USD"
                            }
                        },
                        "unitPrice": {
                            "gross": {
                                "amount": 37.72,
                                "currency": "USD"
                            },
                            "net": {
                                "amount": 35.0,
                                "currency": "USD"
                            },
                            "tax": {
                                "amount": 2.72,
                                "currency": "USD"
                            }
                        },
                        "undiscountedUnitPrice": {
                            "gross": {
                                "amount": 37.71,
                                "currency": "USD"
                            },
                            "net": {
                                "amount": 35.0,
                                "currency": "USD"
                            },
                            "tax": {
                                "amount": 2.71,
                                "currency": "USD"
                            }
                        },
                        "unitDiscount": {
                            "amount": 0.0,
                            "currency": "USD"
                        },
                        "unitDiscountType": "FIXED"
                    }
                ]
            },
            "errors": []
        }
    },
    "extensions": {
        "cost": {
            "requestedQueryCost": 1,
            "maximumAvailable": 50000
        }
    }
}
```

File: saleor/order/migrations/0161_recalculate_undiscounted_prices.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+
+def recalculate_undiscounted_prices_for_order(apps, _schema_editor):
+    def on_migrations_complete(sender=None, **kwargs):
+        recalculate_undiscounted_prices.delay()
+
+    # Make sure that there are OrderLines that needs to be updated
+    # to prevent running unneeded task
+    OrderLine = apps.get_model("order", "OrderLine")
+    if not OrderLine.objects.filter(
```
[the remark this answers is not in the document]
Comment (jakubkuc) [reply saleor/order/migrations/0161_recalculate_undiscounted_prices.py:16]: I did test it, and migrations took around 1 sec and memory usage did not significantly rise. Although it was before some minor changes in this code, I may need to re-test it.

If you think that this check is not needed I can remove it, your arguments make sense to me.

**Comment by aniav** [comment]
What is the status of this? Are we still doing this or should this be closed?

**Comment by maarcingebala** [comment]
@jakubkuc Is working on that, it's still in progress.

**Comment by jakubkuc** [comment]
Closing because it will not be introduced in 3.9, switched to this PR instead #13585
